### DISCUSSION TITLE
Extruding a volume under overhanging surfaces

### DISCUSSION THREAD

**Discussion by JasperBoogers** [pyvista/pyvista#5481] [Q&A]
Hi all! I'm working on a project where I want to generate support structures for a 3D printing setting. Since I'll be doing a lot of rotations and boolean operations on meshes, PyVista seemed the way to go.
So after looking at #2548 and the [surface normals example](https://docs.pyvista.org/version/stable/examples/01-filter/compute-normals.html#computing-surface-normals), I came up with the code below. The basic idea is to extract any overhanging surfaces (z-component of normal vector < 0) and extrude those to a plane below the object, which will give me a rough estimate of the volume of support structures needed if I were to 3D print this.

```import pyvista as pv
import numpy as np
from pyvista import examples

# download mesh and compute normals
mesh = examples.download_cow()
mesh.triangulate(inplace=True).compute_normals(inplace=True)

# extract overhanging surfaces
ids = np.arange(mesh.n_cells)[mesh['Normals'][:, 2] < 0.0]
overhang = mesh.extract_cells(ids).extract_surface()
overhang.triangulate(inplace=True)

# create plane to extrude to
bounds = overhang.bounds
plane = pv.Plane(center=(0, 0, bounds[-2] - 1),
                 i_size=1.1 * (bounds[1] - bounds[0]),
                 j_size=1.1 * (bounds[3] - bounds[2]),
                 direction=(0, 0, 1))

# extrude the overhanging surface to the plane
support = overhang.extrude_trim((0, 0, -1), plane)
support.triangulate(inplace=True).compute_normals(inplace=True)

# extract boundary edges
edges = support.extract_feature_edges(boundary_edges=True, feature_edges=False, manifold_edges=False)

# check
print(f'number of open edges: {support.n_open_edges}')  # 212
print(f'is the support manifold: {support.is_manifold}')  # False

# plot
plot = pv.Plotter()
_ = plot.add_mesh(mesh, opacity=0.5, color='blue')
_ = plot.add_mesh(overhang, opacity=1, color='orange')
_ = plot.add_mesh(support, opacity=0.5, color='green')
_ = plot.add_mesh(edges, color='r', line_width=3)
plot.show()
```

this approach results in the following meshes:
![image](https://github.com/pyvista/pyvista/assets/48677568/307f9433-7362-44b3-8b37-9c54c384a287)

Clearly, the support mesh is open and not clean, so I tried to clean it:
```
# attempt to repair the mesh
import pymeshfix as mf
meshfix = mf.MeshFix(support)
meshfix.repair(joincomp=True, remove_smallest_components=False)
support_repaired = meshfix.mesh.triangulate()

# check
print(f'number of open edges: {support_repaired.n_open_edges}')  # 0
print(f'is the support manifold: {support_repaired.is_manifold}')  # True


```

This results in the following mesh (in purple):
![image](https://github.com/pyvista/pyvista/assets/48677568/7a403d75-4c87-4c90-9417-b97f27841f51)

Although the repair has resulted in a manifold mesh, I have now lost a lot of volume, especially around the legs and head, which is the most interesting part for later stages of the project. Basically I want to subtract the original cow mesh from the support mesh to enable 'support on part' and remove duplicate volumes. This seems quite hard to do without repairing the mesh, but that results in a significant loss of data.

In practice, one would not generate support structures for 3D printing near-vertical faces, so the threshold for extracting overhanging surfaces is lower than 0, say -0.5. This does reduce the number of open edges, but they still remain so it does not solve the problem.

Long story short, I am a bit stuck here. Does anyone have a clue what I can do to retain the support mesh but still be able to perform boolean operations on it? Maybe there's a completely different way?

Thanks in advance for looking into it!

Jasper

**Comment by darikg** [comment]
I have an alternate approach that might work... First, project the entire mesh to the ground plane, then use the 2d geometry library [Shapely](https://shapely.readthedocs.io/en/stable/index.html) to take the union of all the projected triangles

```python
import numpy as np
import pyvista as pv
import shapely

mesh = pv.examples.download_cow().triangulate().compute_normals()
mesh.points[:, 2] -= mesh.points[:, 2].min()  # Make sure it's all above ground
flat = mesh.project_points_to_plane()

merged = shapely.union_all([
    shapely.Polygon(flat.points[tri, :2])
    for tri in flat.regular_faces
])
merged
```

![image](https://github.com/pyvista/pyvista/assets/6875882/7d57a307-3cc4-4cae-9100-872e58dda9f3)

The result has some holes in it so massage it a little so we get a simple polygon. I'm using `buffer` here to first expand the polygon, and then shrink it by the same amount, filling in the holes, and then simplifying it by merging points that are too close together.

```python
cleaned = merged.buffer(0.1).buffer(-0.1).simplify(0.01, preserve_topology=True)
cleaned
```

![image](https://github.com/pyvista/pyvista/assets/6875882/9253c281-4e58-43cb-a70f-2ccc40bce346)

Next, triangulate the polygon so we can uniformly sample points in the interior. I'm using [Triangle](https://rufat.be/triangle/index.html) here because I find VTK's 2d delaunay implementation unreliable.

```python
import triangle

pts = np.array(cleaned.boundary.coords)  # (n_pts, 2)
# the first and last point in this array are the same because the boundary is a closed loop
# We'll ignore the last point so they're all distinct
n_border_pts = len(pts) - 1
border_idx = np.append(np.arange(n_border_pts), 0)  # [0, 1, 2, ..., n_border_pts - 1, 0]

# Edges (0, 1), (1, 2), (2, 3), ..., (n_border_pts - 2, n_border_pts - 1), (n_border_pts - 1, 0)]
edges = np.stack([border_idx[:-1], border_idx[1:]], axis=1)  

# Because triangulate is a wrapper around a C library the syntax is a little weird
# opts='p' means planar straight line graph
# opts='a0.5' means a maximum triangle area of 0.5
tri_data = triangle.triangulate(dict(vertices=pts[:-1], segments=edges), opts='pa0.5')
verts, faces = tri_data['vertices'], tri_data['triangles']
bottom = pv.PolyData.from_regular_faces(points=np.hstack([verts, np.zeros((len(verts), 1))]), faces=faces)
bottom.plot(show_edges=True)
```

![image](https://github.com/pyvista/pyvista/assets/6875882/621e73da-546a-4d0f-97c1-c77cdb056775)

Next, use ray tracing to find the corresponding point in the original mesh vertically above every point in the projected `bottom` mesh. Because of the smoothing/simplification of the 2d polygon, this ray tracing will miss for most of the points near the border, so do the ray tracing on a slightly inflated version of the original.

```python
inflated_pts = mesh.points + 0.1 * mesh.point_data['Normals']
inflated = pv.PolyData(inflated_pts, faces=mesh.faces) 
margin = [0, 0, inflated_pts[:, 2].max() + 1]
top_pts = np.zeros((len(verts), 3))

for i, pt in enumerate(bottom.points):
    upper_pt, _ = inflated.ray_trace(pt - margin, pt + margin, first_point=True)
    if len(upper_pt) > 0:
        if upper_pt[2] < 0:  # Inflated mesh might be below ground
            upper_pt[2] = 0
        top_pts[i] = upper_pt
    else:
        print(f'point {i}: ray trace failed, falling back to closest point')
        top_pts.append(mesh.points[mesh.find_closest_point(pt)])

top = pv.PolyData.from_regular_faces(points=np.array(top_pts), faces=bottom.regular_faces)

p = pv.Plotter()
p.add_mesh(bottom, color='r', show_edges=True)
p.add_mesh(top, color='b', show_edges=True)
p.show()
```

![image](https://github.com/pyvista/pyvista/assets/6875882/07dab41b-dcbc-414a-9290-78a95961305c)

Now, because the top and bottom meshes have identical topologies, we can do a one-to-one mapping between border points to construct the walls:

```python
# Create quadrilateral faces connecting top and bottom borders
all_pts = np.concatenate([bottom.points, top.points])
top_pt_offset = len(bottom.points)

wall_faces = np.stack([
    border_idx[:-1],
    border_idx[1:],
    top_pt_offset + border_idx[1:],
    top_pt_offset + border_idx[:-1],
], axis=1)
walls = pv.PolyData.from_regular_faces(all_pts, wall_faces)
walls.plot(show_edges=True)
```

![image](https://github.com/pyvista/pyvista/assets/6875882/f7aac6e2-ec8f-4354-ab54-94b9be262ab4)

Finally, merge everything together:

```python
all_triangles = np.concatenate([
    bottom.regular_faces,
    top_pt_offset + top.regular_faces,
    walls.triangulate().regular_faces,
])
final = pv.PolyData.from_regular_faces(all_pts, all_triangles)
final.plot(show_edges=True)
```

![image](https://github.com/pyvista/pyvista/assets/6875882/5160d0fb-f9b8-4702-987d-741c0ff34ffe)
